extensions/aws/processors/PutKinesisStream.cpp

```diff
@@ -0,0 +1,164 @@
+    if (!flow_file) { break; }
+    const auto flow_file_size = flow_file->getSize();
+    if (flow_file_size > SINGLE_RECORD_MAX_SIZE) {
+      flow_file->setAttribute(AwsKinesisErrorMessage.name, fmt::format("record too big {}, max allowed {}", flow_file_size, SINGLE_RECORD_MAX_SIZE));
+      session.transfer(flow_file, Failure);
+      logger_->log_error("Failed to publish to kinesis record {} because the size was greater than {} bytes", flow_file->getUUID().to_string(), SINGLE_RECORD_MAX_SIZE);
+      continue;
+    }
+
+    auto stream_name = context.getProperty(AmazonKinesisStreamName.name, flow_file.get());
+    if (!stream_name) {
+      logger_->log_error("Stream name is invalid due to {}", stream_name.error().message());
+      session.transfer(flow_file, Failure);
+      continue;
+    }
+    auto partition_key = context.getProperty(AmazonKinesisStreamPartitionKey.name, flow_file.get())
+        | minifi::utils::valueOrElse([&flow_file]() -> std::string { return flow_file->getUUID().to_string(); });
+
+    stream_batches[*stream_name].flow_files.push_back(std::move(flow_file));
+    stream_batches[*stream_name].batch_size += flow_file_size;
+
+    if (stream_batches[*stream_name].batch_size > batch_data_size_soft_cap_) {
+      break;
+    }
+  }
+
+  std::unique_ptr<Aws::Kinesis::KinesisClient> kinesis_client = getClient(*credentials);
+
+  for (const auto& [stream_name, stream_batch]: stream_batches) {
+    Aws::Kinesis::Model::PutRecordsRequest request;
+    request.SetStreamName(stream_name);
+    Aws::Vector<Aws::Kinesis::Model::PutRecordsRequestEntry> records;
+    for (const auto& flow_file : stream_batch.flow_files) {
+      Aws::Kinesis::Model::PutRecordsRequestEntry entry;
+      const auto partition_key = context.getProperty(AmazonKinesisStreamPartitionKey.name, flow_file.get()) | minifi::utils::valueOrElse([&flow_file] { return flow_file->getUUID().to_string(); });
+      entry.SetPartitionKey(partition_key);
+      const auto [status, buffer] = session.readBuffer(flow_file);
+      Aws::Utils::ByteBuffer aws_buffer(reinterpret_cast<const unsigned char*>(buffer.data()), buffer.size());
+      entry.SetData(aws_buffer);
+      records.push_back(entry);
+    }
+    request.SetRecords(records);
+
+    const auto outcome = kinesis_client->PutRecords(request);
+
+    if (!outcome.IsSuccess()) {
+      for (const auto& flow_file : stream_batch.flow_files) {
+        flow_file->addAttribute(AwsKinesisErrorMessage.name, outcome.GetError().GetMessage());
+        flow_file->addAttribute(AwsKinesisErrorCode.name, std::to_string(static_cast<int>(outcome.GetError().GetErrorType())));
+        session.transfer(flow_file, Failure);
+      }
+    } else {
+      const auto result_records = outcome.GetResult().GetRecords();
+      if (result_records.size() != stream_batch.flow_files.size()) {
+        logger_->log_critical("PutKinesisStream record size mismatch cannot tell which record succeeded and which didnt");
```

Comment:
logging the two sizes could help with troubleshooting; also the stream name could be useful

Reply:
👍 https://github.com/apache/nifi-minifi-cpp/pull/1927/commits/3bf6cb8601b1584cdb9ee1c9a252ee97577699da#diff-7f440937a3123a187fcf52fd1ab61e0d4617345e57a49cbff318988e18cf24d2R133